packages/contracts/contracts/L1/OptimismPortal.sol

```diff
@@ -0,0 +1,158 @@
+    L2OutputOracle public immutable L2_ORACLE;
+
+    /**
+     * @notice Public variable which can be used to read the address of the L2 account which
+     * initated the withdrawal. Can also be used to determine whether or not execution is occuring
+     * downstream of a call to finalizeWithdrawalTransaction().
+     */
+    address public l2Sender = DEFAULT_L2_SENDER;
+
+    /**
+     * @notice A list of withdrawal hashes which have been successfully finalized.
+     * Used for replay protection.
+     */
+    mapping(bytes32 => bool) public finalizedWithdrawals;
+
+    /***************
+     * Constructor *
+     ***************/
+
+    constructor(L2OutputOracle _l2Oracle, uint256 _finalizationPeriod) {
+        L2_ORACLE = _l2Oracle;
+        FINALIZATION_PERIOD = _finalizationPeriod;
+    }
+
+    /**********************
+     * External Functions *
+     **********************/
+
+    /**
+     * @notice Accepts value so that users can send ETH directly to this contract and
+     * have the funds be deposited to their address on L2.
+     * @dev This is intended as a convenience function for EOAs. Contracts should call the
+     * depositTransaction() function directly.
+     */
+    receive() external payable {
+        depositTransaction(msg.sender, msg.value, 30000, false, bytes(""));
+    }
+
+    /**
+     * @notice Finalizes a withdrawal transaction.
+     * @param _nonce Nonce for the provided message.
+     * @param _sender Message sender address on L2.
+     * @param _target Target address on L1.
+     * @param _value ETH to send to the target.
+     * @param _data Data to send to the target.
+     * @param _gasLimit Gas to be forwarded to the target.
+     * @param _timestamp L2 timestamp of the outputRoot.
+     * @param _outputRootProof Inclusion proof of the withdrawer contracts storage root.
+     * @param _withdrawalProof Inclusion proof for the given withdrawal in the withdrawer contract.
+     */
+    function finalizeWithdrawalTransaction(
+        uint256 _nonce,
+        address _sender,
+        address _target,
+        uint256 _value,
+        uint256 _gasLimit,
+        bytes calldata _data,
+        uint256 _timestamp,
+        WithdrawalVerifier.OutputRootProof calldata _outputRootProof,
+        bytes calldata _withdrawalProof
+    ) external {
+        // Check that the timestamp is 7 days old.
+        unchecked {
+            if (block.timestamp < _timestamp + FINALIZATION_PERIOD) {
+                revert NotYetFinal();
+            }
+        }
+
+        // Get the output root.
+        bytes32 outputRoot = L2_ORACLE.getL2Output(_timestamp);
+
+        // Verify that the output root can be generated with the elements in the proof.
+        if (outputRoot != WithdrawalVerifier._deriveOutputRoot(_outputRootProof)) {
+            revert InvalidOutputRootProof();
+        }
+
+        // Verify that the hash of the withdrawal transaction's arguments are included in the
```

Comment:
Computing the `withdrawalHash` could go into the `WithdrawalVerifier` library but not super opinionated